<?xml version="1.0" encoding="utf-8"?>
<feed xmlns="http://www.w3.org/2005/Atom">

  <title><![CDATA[Category: tech | xixitalk's snippet]]></title>
  <link href="http://xixitalk.github.io/blog/categories/tech/atom.xml" rel="self"/>
  <link href="http://xixitalk.github.io/"/>
  <updated>2014-06-21T01:13:44+08:00</updated>
  <id>http://xixitalk.github.io/</id>
  <author>
    <name><![CDATA[xixitalk]]></name>
    
  </author>
  <generator uri="http://octopress.org/">Octopress</generator>

  
  <entry>
    <title type="html"><![CDATA[免费申请S/MIME证书]]></title>
    <link href="http://xixitalk.github.io/blog/2014/06/21/Free-SMIME-Certificate-Key/"/>
    <updated>2014-06-21T08:05:11+08:00</updated>
    <id>http://xixitalk.github.io/blog/2014/06/21/Free-SMIME-Certificate-Key</id>
    <content type="html"><![CDATA[<p>从这里可以申请到免费的S/MIME 邮件证书：<a href="http://www.instantssl.com/ssl-certificate-products/free-email-certificate.html">http://www.instantssl.com/ssl-certificate-products/free-email-certificate.html</a></p>

<p>提交生成之后会把证书发送到邮箱里，在邮箱里点击连接就会把证书安装到电脑系统中（包括windows和Mac OSX)。在电脑系统里的证书管理器里把证书（包含私钥）导出，windows是pfx格式，Mac OSX是p12格式，导出的证书要设置私钥保护密码，iOS的证书安装需要输入这个密码。把这个pfx或者p12格式的证书发邮件到iOS上，点击打开即提示安装，输入之前的保护密码就安装到iOS系统中了。</p>

<!--more-->

<p>在iOS-邮件、通讯录、日历-邮件帐户-高级 启用S/MIME，选择上面安装的证书。iOS自带的mail程序发邮件即可自动进行加密和签名。</p>

<p>在电脑证书管理里导出证书（不包含私钥），windows是cer文件，把这个文件群发出来告诉朋友。别人给你发邮件就用这个证书进行加密。</p>

<p>我的S/MIME证书<a href="http://xixitalk.github.io/static/pub.cer">http://xixitalk.github.io/static/pub.cer</a>,欢迎用S/MIME给我发邮件。</p>

<p>其他参考：<br />
Mac OSX导出证书到iOS参考<a href="http://feinstruktur.com/blog/2011/12/12/using-smime-on-ios-devices.html">http://feinstruktur.com/blog/2011/12/12/using-smime-on-ios-devices.html</a></p>
]]></content>
  </entry>
  
  <entry>
    <title type="html"><![CDATA[curl下载断点传续]]></title>
    <link href="http://xixitalk.github.io/blog/2014/04/13/curl-continue/"/>
    <updated>2014-04-13T01:22:17+08:00</updated>
    <id>http://xixitalk.github.io/blog/2014/04/13/curl-continue</id>
    <content type="html"><![CDATA[<h3 id="curl">curl下载断点传续</h3>

<p><code>
curl -C offset -O http://google.com/p4.mp4 
</code></p>

<p>offset手动输入已经下载的字节数</p>

<p><code>
curl -C - -O http://google.com/p4.mp4 
</code></p>

<p><code>-C -</code>可以自动识别offet续传</p>

<!--more-->

<h3 id="curl-1">curl使用代理</h3>

<p><code>
curl -x 192.168.1.106:8118  http://google.com/p4.mp4 
</code></p>

<p><code>-x</code> 后面跟HTTP代理的host和port</p>

<p><code>
curl  --socks5-hostname 192.168.1.106:9050  http://google.com/p4.mp4
</code></p>

<p><code>--socks5-hostname</code>后面跟socks5代理的host和port</p>

]]></content>
  </entry>
  
  <entry>
    <title type="html"><![CDATA[twitter-python BadStatusLine错误]]></title>
    <link href="http://xixitalk.github.io/blog/2014/04/07/twitter-python-badstatusline/"/>
    <updated>2014-04-07T04:20:25+08:00</updated>
    <id>http://xixitalk.github.io/blog/2014/04/07/twitter-python-badstatusline</id>
    <content type="html"><![CDATA[<p>事情是这样的：我用<a href="https://github.com/bear/python-twitter">twitter-python</a>在本地运行， 抓取一个推号的homeline， 抓取后sleep两分钟再抓取， 不停的循环｡从昨天早上开始， 程序出现如下错误， 且每次都是这个错误｡获得推友收藏列表和fo人的api返回都是正常的，就是REST的api出这种错误。</p>

<!--more-->

<p><code>
Traceback (most recent call last):
  File "hottweetdo.py", line 183, in &lt;module&gt;
    main()
  File "hottweetdo.py", line 126, in main
    status = api2.GetHomeTimeline(since_id=since_id1,count=200)
  File "/media/westdata/read/python-twitter-master/twitter/api.py", line 541, in GetHomeTimeline
    json = self._RequestUrl(url, 'GET', data=parameters)
  File "/media/westdata/read/python-twitter-master/twitter/api.py", line 3383, in _RequestUrl
    timeout=self._requests_timeout
  File "/usr/local/lib/python2.7/dist-packages/requests/api.py", line 55, in get
    return request('get', url, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/api.py", line 44, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 383, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/sessions.py", line 486, in send
    r = adapter.send(request, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/requests/adapters.py", line 382, in send
    raise ConnectionError(e)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='api.twitter.com', port=443): Max retries exceeded with url: /1.1/statuses/home_timeline.json?count=200 (Caused by &lt;class 'httplib.BadStatusLine'&gt;: '')
</code></p>

<p>代码很简单，就是用四个key生成一个api对象，然后直接调用<code>GetHomeTimeline</code></p>

<p>我google之后， 看到很多类似的分析， 但是我都尝试了， 还是没有解决｡尝试罗列如下｡</p>

<h3 id="bad-proxy-server">bad proxy server</h3>
<p><a href="https://stackoverflow.com/questions/18478013/python-requests-exceptions-connectionerror-max-retries-exceeded-witth-url">有人说是bad proxy server</a>,程序确实使用代理了，代理是shadowsocks+polipo， 代理设置在环境变量HTTP_PROXY和HTTPS_PROXY｡我尝试了不同地区的代理， 甚至切换了不同卖家的代理， 依然出错｡</p>

<h3 id="dns">DNS问题</h3>
<p>尝试了不同的DNS server不起作用</p>

<h3 id="httplib2-08-077">httplib2 0.8的问题， 使用0.7.7</h3>
<p>不起作用</p>

<h3 id="twitter-api">twitter API的问题</h3>
<p>twitter-python的一个<a href="https://github.com/bear/python-twitter/issues/131">issues解答</a>说是twitter API诡异的问题｡问题是我只能等待吗？</p>

<h3 id="maxretries5">设置maxretries为5</h3>
<p><a href="https://stackoverflow.com/questions/15431044/can-i-set-max-retries-for-requests-request/15431343#15431343">https://stackoverflow.com/questions/15431044/can-i-set-max-retries-for-requests-request/15431343#15431343</a> 不起作用</p>

<h3 id="requestssession-keep-alivefalse">设置requests的session keep-alive为False</h3>
<p>不起作用</p>

<p>请python高手或者twitter api高手指点迷津，可以直接在下面留言，或者推上@我。</p>

<h2 id="section">[更新后记]</h2>

<p>目前解决的办法：1、重新新建一个API，用新的consumer_key和access token；2、更换代理，换用tor+polipo，tor的IP变化比较频繁。经过上面操作，错误暂时不再出现了。</p>

<p>问题可能原因：应该是twitter服务器认为API滥用了，当get提交给服务器后，twitter服务器只返回了ACK，接下来没有返回任何数据，所以httplib判断为BadStatusLine，因为没有数据，所以连status号都没有。</p>

<h2 id="section-1">[再次更新后记]</h2>
<p>每天生成HTMLpush到github后自动发推又出现BadStatusLine了，所幸不是<code>GetHomeTimeline</code>。</p>

<p><code>
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='api.twitter.com', port=443): Max retries exceeded with url: /1.1/statuses/update.json (Caused by &lt;class 'httplib.BadStatusLine'&gt;: '')
</code></p>
]]></content>
  </entry>
  
  <entry>
    <title type="html"><![CDATA[python里用smtplib发邮件使用代理]]></title>
    <link href="http://xixitalk.github.io/blog/2014/04/04/smtplib-with-proxy/"/>
    <updated>2014-04-04T17:51:49+08:00</updated>
    <id>http://xixitalk.github.io/blog/2014/04/04/smtplib-with-proxy</id>
    <content type="html"><![CDATA[<p>关键词：smtplib mail proxy</p>

<p>树莓派上一个cron python程序直接用smtplib发送gmail邮件昨天出现问题了，错误是：socket.error: [Errno 97] Address family not supported by protocol，之前都是好好的，我用的smtp.gmail.com，端口是587。</p>

<p>我搜索了一圈，觉得原因是国家防火墙搞得鬼，如果smtplib添加代理就可以避免这种情况。搜索了很多方法都觉得不好使，突然想到以前为google app engine的appcfg.py使用SocksiPy添加代理的方法，并且SocksiPy现在被包含在httplib2的库里了，这样就更方便了。</p>

<!--more-->

<p>在使用smtplib操作文件前部添加：</p>

<pre><code>from httplib2 import socks
import socket
socket.socket = socks.socksocket
socks.setdefaultproxy(socks.PROXY_TYPE_SOCKS5,"192.168.1.106",1081)
</code></pre>

<p>经测试邮件发送正常，待后续观察。</p>

<p>socks支持SOCKS4、SOCKS5和HTTP代理，上述代码自行替换之。</p>
]]></content>
  </entry>
  
  <entry>
    <title type="html"><![CDATA[搭建局域网跨路由网络]]></title>
    <link href="http://xixitalk.github.io/blog/2013/06/22/routing-cross/"/>
    <updated>2013-06-22T17:34:57+08:00</updated>
    <id>http://xixitalk.github.io/blog/2013/06/22/routing-cross</id>
    <content type="html"><![CDATA[<p>场景：两个路由器每个都组建一个局域网，每个局域网里的电脑想访问另一个局域网的电脑。</p>

<!--more-->

<p><a href='http://www.flickr.com/photos/xixitalk/9108420474/'><img src='http://farm6.static.flickr.com/5544/9108420474_421bdc9dc4_b.jpg' title='跨路由器访问设置'/></a></p>

]]></content>
  </entry>
  
</feed>
